patches/server/0933-Apply-offset-when-riding-entity.patch

```diff
@@ -0,0 +1,29 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
+Date: Mon, 8 Aug 2022 13:46:23 +0200
+Subject: [PATCH] Apply offset when riding entity
+
+
+diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+index dbefefc123e3cb38a95bd321d74c384bc087cfba..d04344d0559ddbc20ea8f1f1465b56249ddd5f84 100644
+--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
++++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+@@ -692,7 +692,8 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
+                 Location curPos = this.getCraftPlayer().getLocation(); // Spigot
+ 
+                 entity.absMoveTo(d3, d4, d5, f, f1);
+-                this.player.absMoveTo(d3, d4, d5, this.player.getYRot(), this.player.getXRot()); // CraftBukkit
++                double riderOffset = entity.getPassengersRidingOffset() + this.player.getMyRidingOffset(); // Paper
++                this.player.absMoveTo(d3, d4 + riderOffset, d5, this.player.getYRot(), this.player.getXRot()); // CraftBukkit // Paper
```

Comment:
Looking at this, do we even need to sync the players position like this? I really think we should look more indepth if this is actually needed. Because obviously it's desyncing the entire player, but it might not even be needed in the first place.

The only possible place I could think of is
```java
// Check to see if the Players Location has some how changed during the call of the event.
                        // This can happen due to a plugin teleporting the player instead of using .setTo()
                        if (!from.equals(this.getCraftPlayer().getLocation()) && this.justTeleported) {
                            this.justTeleported = false;
                            return;
                        }
```

This PlayerMoveEvent that is being fired here just seems wrong tbh, as it's not the player that's moving it's the entity. And bukkit defo tries to shove the player in the vehicle inorder to pretend this issue doesn't exist.
```java
Location from = new Location(player.getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location.
                Location to = player.getLocation().clone(); // Start off the To location as the Players current location.

                // If the packet contains movement information then we update the To location with the correct XYZ.
                to.setX(packet.getX());
                to.setY(packet.getY());
                to.setZ(packet.getZ());


                // If the packet contains look information then we update the To location with the correct Yaw & Pitch.
                to.setYaw(packet.getYRot());
                to.setPitch(packet.getXRot());
```

What we need to do is "estimate" where the player is going to be moving TO, as this location is the vehicle location and not the players.
Take a look at Entity#positionRider(Entity passenger, Entity.MoveFunction positionUpdater).
What we can do is make our own MoveFunction which updates the TO location based on the offset of the player.

This should allow:
1. Not needing to move the player at all
2. Correctly estimate the new PlayerMoveEvent location